File: nailgun/nailgun/expression/objects.py
#    Copyright 2014 Mirantis, Inc.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.


class ScalarWrapper(object):
    def __init__(self, value):
        self.value = value

    def evaluate(self):
        return self.value

    def __call__(self):
        return self.value


class SubexpressionWrapper(object):
    def __init__(self, subexpression):
        self.subexpression = subexpression

    def evaluate(self):
        self.value = self.subexpression()
        return self.value

    def __call__(self):
        self.evaluate()
        return self.value


class ModelPath(object):
    def __init__(self, path):
        path_parts = path.split(':')
        if len(path_parts) == 1:
            self.model_name = 'default'
            self.attribute = path_parts[0]
        else:
            self.model_name = path_parts[0]
            self.attribute = path_parts[1]

    def set_model(self, models):
        if self.model_name not in models:
            raise KeyError('No model with name "{0}" defined'.format(
                self.model_name))
        self.model = models[self.model_name]

    def get_value(self):
        def get_attribute_value(model, path):
            value = model[path.pop(0)]
            return get_attribute_value(value, path) if len(path) else value
        return get_attribute_value(self.model, self.attribute.split('.'))


class ModelPathWrapper(object):
    def __init__(self, path, expression):
        self.path = path
        self.model_path = ModelPath(path)
        self.expression = expression

    def evaluate(self):
        self.model_path.set_model(self.expression.models)
        result = None
        try:
            result = self.model_path.get_value()
        except (KeyError, AttributeError):
            if self.expression.strict:
                raise TypeError(
                    'Value of {0} is undefined. Set options.strict'
                    ' to false to allow undefined values.'.format(self.path))
        self.value = result
        return self.model_path

    def __call__(self):
        self.evaluate()
        return self.value

File: nailgun/static/expression/objects.js
/*
 * Copyright 2014 Mirantis, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
**/
import _ from 'underscore';
import ExpressionParser from 'expression/parser';

    class ModelPath {
        constructor(path) {
            var pathParts = path.split(':');
            if (_.isUndefined(pathParts[1])) {
                this.modelName = 'default';
                this.attribute = pathParts[0];
            } else {
                this.modelName = pathParts[0];
                this.attribute = pathParts[1];
            }
            return this;
        }

        setModel(models, extraModels = {}) {
            this.model = extraModels[this.modelName] || models[this.modelName];
            if (!this.model) {
                throw new Error('No model with name "' + this.modelName + '" defined');
            }
            return this;
        }

        get(options) {
            return this.model.get(this.attribute, options);
        }

        set(value, options) {
            return this.model.set(this.attribute, value, options);
        }

        change(callback, context) {
            return this.model.on('change:' + this.attribute, callback, context);
        }
    }

    class ScalarWrapper {
        constructor(value) {
            this.value = value;
        }

        evaluate() {
            return this.value;
        }

        getValue() {
            return this.value;
        }
    }

    class SubexpressionWrapper {
        constructor(subexpression) {
            this.subexpression = subexpression;
        }

        evaluate() {
            return this.subexpression();
        }

        getValue() {
            return this.subexpression();
        }
    }

    class ModelPathWrapper {
        constructor(modelPathText) {
            this.modelPath = new ModelPath(modelPathText);
            this.modelPathText = modelPathText;
        }

        evaluate() {
            var expression = ExpressionParser.yy.expression;
            this.modelPath.setModel(expression.models, expression.extraModels);
            var result = this.modelPath.get();
            if (_.isUndefined(result)) {
                if (expression.strict) {
                    throw new TypeError('Value of ' + this.modelPathText + ' is undefined. Set options.strict to false to allow undefined values.');
                }
                result = null;
            }
            this.lastResult = result;
            expression.modelPaths[this.modelPathText] = this.modelPath;
            return this.modelPath;
        }

        getValue() {
            this.evaluate();
            return this.lastResult;
        }
    }

    export {ScalarWrapper, SubexpressionWrapper, ModelPathWrapper, ModelPath};
